proposals/devcontainer-templates-distribution.md

```diff
@@ -0,0 +1,120 @@
+# Distribution and Discovery of Dev Container Templates
+
+This specification defines a pattern where community members and organizations can author and self-publish [Dev Container Templates](./devcontainer-templates.md). 
+
+Goals include:
+
+- For Template authors, create a "self-service" way to publish a Template, either publicly or privately, that is not centrally controlled.
+- Provide the ability to standardize publishing such that supporting tools may implement their own mechanism to aid Template discoverability as they see fit.
+
+## Source code
+
+Templates source code is stored in a git repository.
+
+For ease of authorship and maintenance, [1..n] Templates can share a single git repository. This set of Templates is referred to as a "collection," and will share the same [`devcontainer-collection.json`](#devcontainer-collection.json) file and "namespace" (eg. `<owner>/<repo>`).
+
+> **Note:** Templates and [Features](./devcontainer-features.md) should be placed in different git repositories. 
+
+Source code for the set follows the example file structure below:
+
+```
+.
+├── README.md
+├── src
+│   ├── dotnet
+│   │   ├── devcontainer-template.json
+│   │   ├── .devcontainer.json
+│   │   ├── ...
+|   ├
+│   ├── docker-from-docker
+│   │   ├── devcontainer-template.json
+│   │   ├── .devcontainer
+│   │       ├── devcontainer.json
+│   │       ├── Dockerfile
+│   │       └── ...
+│   │   ├── ...
+|   ├
+│   ├── go-postgres
+│   │   ├── devcontainer-template.json
+│   │   ├── .devcontainer
+│   │       ├── devcontainer.json
+│   │       ├── docker-compose.yml
+│   │       ├── Dockerfile
+│   │       └── ...
+│   │   ├── ...
+```
+
+...where `src` is a directory containing a sub-folder with the name of the Template (e.g. `src/dotnet` or `src/docker-from-docker`) with at least a file named `devcontainer-template.json` that contains the Template metadata, and a `.devcontainer.json` (or `.devcontainer/devcontainer.json`) that the supporting tools will drop into an existing project or folder.
+
+Each sub-directory should be named such that it matches the `id` field of the `devcontainer-template.json`.  Other files can also be included in the Templates's sub-directory, and will be included during the [packaging step](#packaging) alongside the two required files.  Any files that are not part of the Templates's sub-directory (e.g. outside of `src/dotnet`) will not included in the [packaging step](#packaging).
+
+## Versioning
+
+Each Template is published with only the `latest` tag. Tooling that handles releasing Templates will republish the `latest` tag every time a new release is created.
```

Comment:
Hmmm. Not sure we landed on only publishing latest.  I think what we figured out is that we'd want versioning, but that we also need to factor in the schema version. This would allow us to do automated verion selection based on the schema much like VS Code extensions do today down the road. But you'd need to have some concept of versioning to make that work. Thoughts @chrmarti @alexdima?

Reply:
#22 would reference only a devcontainer.json as I understand it, not an entire template.

My take-away on template versioning was that it would not solve the schema version compatibility (the same way the package.json's version property doesn't solve the Node.js version compatibility) and that we would look into adding a `"schemaVersion"` (or similar) property (similar to the package.json's engines property) in a separate proposal to not overload this one.

Reply:
I am quite confident that when we expand #22 to supporting remote devcontainer.jsons, customers will want an entire template since this is generally how the request has come across.  We limited #22 to local to start because this proposal was not in place. The request we've heard often is to have the ability to reference a remote template so that when the template updates with non-breaking changes, downstream repositories get the updates without modification. To achieve this, we'd need to support referencing MAJOR or MAJOR.MINOR versions of templates like we do with features.

Also, with the schema version property, I agree that we do not want to overload this proposal, but fundamentally that will require being able to go back to a previous version of a template.

In either case, including the same versioning semantics not only keeps things consistent with features, but lets clients implement as they see fit and we don't have to retroactively go back and update this proposal and how anyone that is already publishing templates is doing so.

We can omit this from the proposal here - but given what we already know, and how features is set up, why would we do that?

Reply:
Writing down my thoughts for semver vs always publishing `latest` (mostly reiterating what Chuck already mentioned)
- Using semantic versioning will make publishing consistent with Features - it will avoid questions from the community of why one logic for Features and other for Templates.
- Given current consumption of definitions/templates by the Remote Containers UX, I agree versioning does not make complete sense and it doesn't sound helpful. Hence, for UX we can always pin the `latest` tag. However, there is a possibility that the community could develop some tool or service which might require pinning specific versions. So it will be nice to have (just in case)
    - Given our current set of maintained Templates, we might not think that we may add breaking changes to them. However, given that the Templates will be published and distributed by community as well, we can't be sure what pattern they might use. 

To sum up, I believe that having semver may not have useful scenarios currently, but it could be nice to have to provide flexibility and adaptability for future scenarios. Also, I don't see a downside of publishing semantic versions, unless I am missing something super obvious 😄

Reply:
I agree it might be useful in the future, I'm fine with adding it now.